fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MinimumTupleSizeKeyChecker.java

```diff
@@ -0,0 +1,185 @@
+        private final int maxTuplePrefixCount;
+
+        /**
+         * Create a new checked subspace.
+         * @param prefix key prefix to which these checks apply
+         * @param dryRun whether to throw or just log problems
+         * @param checkReads whether to also check read operations
+         * @param minTupleSize minimum size (number of elements) of tuple-decoded keys
+         * @param maxTuplePrefixCount maximum number of different prefixes
+         */
+        public CheckedSubspace(@Nonnull byte[] prefix,
+                               boolean dryRun, boolean checkReads,
+                               int minTupleSize, int maxTuplePrefixCount) {
+            this.prefix = prefix;
+            this.dryRun = dryRun;
+            this.checkReads = checkReads;
+            this.minTupleSize = minTupleSize;
+            this.maxTuplePrefixCount = maxTuplePrefixCount;
+        }
+    }
+
+    /**
+     * Create a new checker.
+     * @param checks ordered list of checks to match and apply
+     */
+    public MinimumTupleSizeKeyChecker(@Nonnull List<CheckedSubspace> checks) {
+        this.checks = checks;
+    }
+
+    @Override
+    public void checkKey(final byte[] key, final boolean mutation) {
+        final CheckedSubspace checkedSubspace = findCheckedSubspace(key);
+        if (checkedSubspace == null) {
+            return;
+        }
+        if (!checkedSubspace.checkReads && !mutation) {
+            return;
+        }
+        checkKeyLength(checkedSubspace, key);
+    }
+
+    @Override
+    public void checkKeyRange(final byte[] keyBegin, final byte[] keyEnd, final boolean mutation) {
+        final CheckedSubspace checkedSubspace = findCheckedSubspace(keyBegin);
+        if (checkedSubspace == null) {
+            return;
+        }
+        if (!checkedSubspace.checkReads && !mutation) {
+            return;
+        }
+        final byte[] prefix = checkKeyLength(checkedSubspace, keyBegin);
+        if (prefix != null) {
+            if (!keyHasPrefix(keyEnd, prefix)) {
+                checkFails(checkedSubspace, "key range not limited to single subspace",
+                        "keyBegin", ByteArrayUtil2.loggable(keyBegin),
+                        "keyEnd", ByteArrayUtil2.loggable(keyEnd));
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        for (CheckedSubspace checkedSubspace : checks) {
+            if (checkedSubspace.dryRun) {
+                checkPrefixesCount(checkedSubspace);
+            }
+        }
+    }
+
+    @Nullable
+    private CheckedSubspace findCheckedSubspace(@Nonnull byte[] key) {
+        for (CheckedSubspace checkedSubspace : checks) {
+            if (keyHasPrefix(key, checkedSubspace.prefix)) {
+                return checkedSubspace;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private synchronized byte[] checkKeyLength(@Nonnull CheckedSubspace checkedSubspace, final byte[] key) {
```

Comment:
I'm trying to make sure I understand this check. The idea is that:

1. The key should begin with a prefix that is some `Tuple` beginning with `minTupleSize` elements
2. If the key checker has already seen a key in that prefix, it returns the pre-existing byte array
3. Otherwise, it records the new prefix

So, at the end, you get a full set of all prefixes seen that where at least `minTupleSize` in size.

I guess that makes sense, though I guess there's a question as to whether you'd want this to return `Tuple`s or `byte[]`.

I suppose a slightly different way would be to have a `KeyChecker` that took a `KeySpace`, and then this could find the `KeySpacePath` leaf that the key (range) was in, and that might tell you similar information. You could also limit that to a given number of path elements, if you really wanted to.

One place I could see this breaking down is the directory layer, as the default directory layer begins at a prefix that is not Tuple-encodable (namely `\xfe`). That would fall over in both a record-key-space-path approach or this approach.

Reply:
Yes. I think a special `KeyChecker` is going to be needed for the directory layer.